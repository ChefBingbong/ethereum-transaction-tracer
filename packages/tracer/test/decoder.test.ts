import type { EventTopic } from '@evm-tt/utils'
import {
  encodeFunctionData,
  encodeFunctionResult,
  erc20Abi,
  getAbiItem,
  stringify,
  zeroAddress,
} from 'viem'
import { describe, expect, it } from 'vitest'
import {
  safeDecodeCallData,
  safeDecodeCallResult,
  safeDecodeCallRevert,
  safeDecodeEvent,
} from '../src'
import { Permit2 } from './abis/Permit2.abi'

describe('Decoder integration', () => {
  it('safeDecodeCallResult: decodes function return', () => {
    const encoded = encodeFunctionResult({
      abi: erc20Abi,
      functionName: 'balanceOf',
      result: 100000n,
    })

    const balanceOfAbiItem = getAbiItem({ abi: erc20Abi, name: 'balanceOf' })

    const [err, res] = safeDecodeCallResult(balanceOfAbiItem, encoded)

    expect(err).toBeUndefined()
    expect(res).toBe('100000')
  })

  it('safeDecodeCallData: decodes function data', () => {
    const encoded = encodeFunctionData({
      abi: erc20Abi,
      functionName: 'balanceOf',
      args: [zeroAddress],
    })

    const balanceOfAbiItem = getAbiItem({ abi: erc20Abi, name: 'balanceOf' })

    const [err, res] = safeDecodeCallData([balanceOfAbiItem], encoded)

    expect(err).toBeUndefined()

    expect(res).toHaveProperty('functionName', 'balanceOf')
    expect(res).toHaveProperty('args', [['account', zeroAddress]])
    expect(res).toMatchSnapshot('safeDecodeCallData')
  })

  it('safeDecodeRevert: decodes revert data', () => {
    const node = {
      output: '0x756688fe',
    } as const

    const [err, res] = safeDecodeCallRevert(Permit2, node.output)

    expect(err).toBeUndefined()

    expect(res).toHaveProperty('errorName', 'InvalidNonce')
    expect(res).toMatchSnapshot('safeDecodeRevert')
  })
  it('safeDecodeEvent: decodes event name and args', () => {
    const data =
      '0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff0000000000000000000000000000000000000000000000000000000068d2ed4e0000000000000000000000000000000000000000000000000000000000000002'
    const topics: EventTopic = [
      '0xc6a377bfc4eb120024a8ac08eef205be16b817020812c73223e81d1bdb9708ec',
      '0x000000000000000000000000da8a8833e938192781ade161d4b46c4973a40402',
      '0x000000000000000000000000a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
      '0x00000000000000000000000066a9893cc07d91d95644aedd05d03f95e1dba8af',
    ]

    const eventAbi = getAbiItem({ abi: Permit2, name: 'Permit' })
    const [err, res] = safeDecodeEvent(eventAbi, topics, data)

    expect(err).toBeUndefined()

    expect(res).toHaveProperty('eventName', 'Permit')

    expect(stringify(res?.args)).toEqual(stringify(res?.args))
    expect(res).toMatchSnapshot('safeDecodeEvent')
  })
})
