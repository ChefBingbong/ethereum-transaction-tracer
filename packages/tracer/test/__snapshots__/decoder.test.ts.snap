// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Decoder integration > safeDecodeCallData: decodes function data > safeDecodeCallData 1`] = `
{
  "args": [
    [
      "account",
      "0x0000000000000000000000000000000000000000",
    ],
  ],
  "functionName": "balanceOf",
  "item": {
    "inputs": [
      {
        "name": "account",
        "type": "address",
      },
    ],
    "name": "balanceOf",
    "outputs": [
      {
        "type": "uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
}
`;

exports[`Decoder integration > safeDecodeEvent: decodes event name and args > safeDecodeEvent 1`] = `
{
  "args": {
    "amount": 1461501637330902918203684832716283019655932542975n,
    "expiration": 1758653774,
    "nonce": 2,
    "owner": "0xda8A8833E938192781AdE161d4b46c4973A40402",
    "spender": "0x66a9893cC07D91D95644AEDD05D03f95e1dBA8Af",
    "token": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
  },
  "eventName": "Permit",
}
`;

exports[`Decoder integration > safeDecodeRevert: decodes revert data > safeDecodeRevert 1`] = `
{
  "abiItem": {
    "inputs": [],
    "name": "InvalidNonce",
    "type": "error",
  },
  "args": undefined,
  "errorName": "InvalidNonce",
}
`;

exports[`Decoder integration safeDecodeCallData: decodes function data: safeDecodeCallData 1`] = `
{
  "args": [
    [
      "account",
      "0x0000000000000000000000000000000000000000",
    ],
  ],
  "functionName": "balanceOf",
  "item": {
    "inputs": [
      {
        "name": "account",
        "type": "address",
      },
    ],
    "name": "balanceOf",
    "outputs": [
      {
        "type": "uint256",
      },
    ],
    "stateMutability": "view",
    "type": "function",
  },
}
`;

exports[`Decoder integration safeDecodeEvent: decodes event name and args: safeDecodeEvent 1`] = `
{
  "args": {
    "amount": 1461501637330902918203684832716283019655932542975n,
    "expiration": 1758653774,
    "nonce": 2,
    "owner": "0xda8A8833E938192781AdE161d4b46c4973A40402",
    "spender": "0x66a9893cC07D91D95644AEDD05D03f95e1dBA8Af",
    "token": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
  },
  "eventName": "Permit",
}
`;

exports[`Decoder integration safeDecodeRevert: decodes revert data: safeDecodeRevert 1`] = `
{
  "abiItem": {
    "inputs": [],
    "name": "InvalidNonce",
    "type": "error",
  },
  "args": undefined,
  "errorName": "InvalidNonce",
}
`;
